src_Core/Core/CoreW.bsv

```diff
@@ -423,78 +484,118 @@ module mkCoreW #(Reset dm_power_on_reset)
    endrule
 
    // ================================================================
-   // INTERFACE
-
-   // ----------------------------------------------------------------
-   // Debugging: set core's verbosity, htif addrs
+   // Connect external debug module interface
 
-   method Action  set_verbosity (Bit #(4)  verbosity, Bit #(64)  logdelay);
-      // Warning: ignoring logdelay
-      proc.set_verbosity (verbosity);
-   endmethod
+   let dbgShim <- mkAXI4LiteShim (reset_by porReset);
 
-   // ----------------------------------------------------------------
-   // Start
-
-   method Action start (Bool is_running, Bit #(64) tohost_addr, Bit #(64) fromhost_addr);
-      plic.set_addr_map (zeroExtend (soc_map.m_plic_addr_range.base),
-                         zeroExtend (rangeTop(soc_map.m_plic_addr_range)));
-
-      let pc = soc_map_struct.pc_reset_value;
-      proc.start (is_running, pc, tohost_addr, fromhost_addr);
-
-`ifdef INCLUDE_GDB_CONTROL
-      // Save for potential future use by rl_dm_harts_reset
-      rg_tohost_addr   <= tohost_addr;
-      rg_fromhost_addr <= fromhost_addr;
-`endif
+   rule rl_debug_module_read_req;
+      let arFlit <- get (dbgShim.master.ar);
+      debug_module.dmi.read_addr (truncate (arFlit.araddr >> 2));
+   endrule
+   rule rl_debug_module_read_rsp;
+      let x <- debug_module.dmi.read_data;
+      dbgShim.master.r.put(AXI4Lite_RFlit { rdata: x, rresp: OKAY, ruser: ?});
+   endrule
+   rule rl_debug_module_write_req;
+      let awFlit <- get (dbgShim.master.aw);
+      let wFlit <- get (dbgShim.master.w);
+      dbgShim.master.b.put(defaultValue);
+      debug_module.dmi.write (truncate (awFlit.awaddr >> 2), wFlit.wdata);
+   endrule
 
-      $display ("%0d: %m.method start: proc.start (pc %0h, tohostAddr %0h, fromhostAddr %0h)",
-                cur_cycle, pc, tohost_addr, fromhost_addr);
-   endmethod
+   let fromDbgReset <- mkPulseWire (reset_by porReset);
+   Reg #(UInt #(8)) ndm_reset_delay <- mkReg (0, reset_by porReset);
+   Reg #(Bool) ndm_reset_restart_running <- mkReg (False, reset_by porReset);
+   rule rl_debug_module_send_reset (ndm_reset_delay == 0);
+      let restartRunning <- debug_module.ndm_reset_client.request.get;
+      ndm_reset_delay <= 110;
```

Comment:
Where is 110 coming from as reset delay?

Reply:
from the P3 wrapper.